dataclasses_avroschema/fields/fields.py

```diff
@@ -324,6 +325,116 @@ def fake(self) -> typing.Any:
         return field.fake()
 
 
+@dataclasses.dataclass
+class LiteralField(Field):
+    """
+    Supports typing.Literal. Note that typing.Literal[v1, v2, v3] is
+    Python's shorthand for typing.Union[typing.Literal[v1], typing.Literal[v2], typing.Literal[v3]]
+    and will be treated as such.
+
+    Note the following behavior when a string and matching enum are both
+    included in Literal:
+        class MyEnum(enum.Enum):
+            ONE = "one"
+
+        @dataclasses.dataclass
+        class MyModel(AvroModel):
+            field: typing.Literal["one", MyEnum.ONE]
+
+        MyModel.parse_obj({"field": "one"})
+
+    will always return:
+
+        # MyModel(field=<MyEnum.ONE: 'one'>)
+
+    This is because typing.Literal["one", MyEnum.ONE] is translated to
+    typing.Union[typing.Literal["one"], typing.Literal[MyEnum.ONE]]
+    and for matching a value to a union of literals, enums are prioritized
+    over other types.
+    """
+
+    allowed_values: typing.Set[typing.Any] = dataclasses.field(default_factory=set)
+    # This isn't actually optional, but Field has optional fields, so we have to provide a default
+    avro_field: typing.Optional[Field] = None
+
+    def __post_init__(self) -> None:
+        """
+        Derives Avro schema type[s] and validation requirements
+        """
+        args = get_args(self.type)
+        args_length = len(args)
+        if args_length > 1:
+            # This field is of the form typing.Literal[v1, v2, v3], which is a union of Literals
+            native_type = typing.Union[tuple(typing.Literal[a] for a in args)]  # type: ignore
+        else:
+            native_type = type(args[0])
+
+        self.allowed_values = set(args)
+        self.avro_field = AvroField(
+            name=self.name,
+            native_type=native_type,
+            parent=self.parent,
+            model_metadata=self.model_metadata,
+        )
+
+    def get_avro_type(self) -> typing.Any:
+        return self.avro_field.get_avro_type()  # type: ignore
+
+    def validate_value(self, value: typing.Any) -> bool:
+        """
+        Validates values assigned to Literal fields
+        """
+        if value not in self.allowed_values:
+            raise FieldValueError(field_name=self.name, field_type=self.type, field_value=value)
+
+        return True
+
+    def validate_default(self, default: typing.Any) -> bool:
+        return self.validate_value(default)
+
+    def get_dacite_typehook_transformer(self) -> typing.Callable[[typing.Any], typing.Any]:
```

Comment:
So I added this one because `dacite` fails to convert a raw value back to its enumeration for `typing.Literal`. For instance, it handles this fine
```py
from dataclasses import dataclass
from typing import Literal
import enum
import dacite

@dataclass
class Model:
    version: Literal['v1']

print(dacite.from_dict(Model, {'version': 'v1'}))
# Model(version='v1')
```
but not this
```py
from dataclasses import dataclass
from typing import Literal
import enum
import dacite

class Versions(enum.Enum):
    v1 = 'v1'
    v2 = 'v2'

@dataclass
class Model:
    version: Literal[Versions.v1]

print(dacite.from_dict(Model, {'version': 'v1'}))
# dacite.exceptions.WrongTypeError: wrong value type for field "version" - should be "Literal" instead of value "v1" of type "str"
```

[An issue](https://github.com/konradhalas/dacite/issues/109) was opened against it 3 years ago, but the library has yet to support it, so this `typehook` provides that support. We could make users provide this manually, but then every user that wants to use `typing.Literal` with `enum`s would have to define and provide this typehook. It feels complicated enough that it would discourage folks from using literals with enums. Thoughts?

Reply:
Okay, I'm with you. So there's no guarantee with `AvroModel` dataclasses that the data in or out is correct - it's just providing the schema and interface for ser/deser. If users want type/value constraints then they need to rely on `pydantic` via `AvroBaseModel`. I'll back this out as well 👍

Reply:
Exactly. There is a guarantee only when data is send/receive from  a stream  (kafka, redis, etc). An instance can be `serialized/deserialize` with the schema that it represents only if the data is correct. If the data is incorrect then `fastavro` validation will raise an exception. This design in on purpose in order to allow the 3 flavors:

1. dataclasses user (no validation)
2. dacite users (optional validation)
3. pydantic users (with validation)

Reply:
Makes total sense 👌